src/System.Management.Automation/engine/NativeCommandProcessor.cs

```diff
@@ -1352,31 +1595,46 @@ internal ProcessInputWriter(InternalCommand command)
         {
             Dbg.Assert(command != null, "Caller should validate the parameter");
             _command = command;
+
+            _pipeline = ScriptBlock.Create("Out-String -Stream").GetSteppablePipeline();
+            _pipeline.Begin(true);
         }
 
         #endregion constructor
 
-        /// <summary>
-        /// Input is collected in this list
-        /// </summary>
-        private ArrayList _inputList = new ArrayList();
+        private SteppablePipeline _pipeline;
+
         /// <summary>
         /// Add an object to write to process
         /// </summary>
         /// <param name="input"></param>
         internal void Add(object input)
         {
-            _inputList.Add(input);
-        }
+            if (_stopping || _streamWriter == null)
+            {
+                // if _streamWriter is already null, then we already called Done()
+                // so we should just discard the input.
+                return;
+            }
 
-        /// <summary>
-        /// Count of object in inputlist
-        /// </summary>
-        internal int Count
-        {
-            get
+            Array formattedObjects = _pipeline.Process(input);
+            foreach (var item in formattedObjects)
             {
-                return _inputList.Count;
+                string line = PSObject.ToStringParser(_command.Context, item);
+                // if process is already finished and we are trying to write something to it,
+                // we will get IOException
+                try
+                {
+                    _streamWriter.WriteLine(line);
+                }
+                catch (IOException)
+                {
+                    // we are assuming that process is already finished
+                    // we should just stop processing at this point
```

Comment:
Minor: may be we can avoid this exception by using the overload `StreamWriter(Stream stream, Encoding encoding, int bufferSize, bool leaveOpen);`. The writer will not try to dispose the stream when `leaveOpen = true`. Let the process object handle closing the stream.